orbit/pkg/scripts/scripts.go

```diff
@@ -0,0 +1,169 @@
+// Package scripts implements support to execute scripts on the host when
+// requested by the Fleet server.
+package scripts
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+)
+
+const scriptExecTimeout = 30 * time.Second
+
+// Client defines the methods required for the API requests to the server. The
+// fleet.OrbitClient type satisfies this interface.
+type Client interface {
+	GetHostScript(execID string) (*fleet.HostScriptResult, error)
+	SaveHostScriptResult(result *fleet.HostScriptResultPayload) error
+}
+
+// Runner is the type that processes scripts to execute, taking care of
+// retrieving each script, saving it in a temporary directory, executing it and
+// saving the results.
+type Runner struct {
+	Client                 Client
+	ScriptExecutionEnabled bool
+
+	// tempDirFn is the function to call to get the temporary directory to use,
+	// inside of which the script-specific subdirectories will be created. If nil,
+	// the user's temp dir will be used (can be set to t.TempDir in tests).
+	tempDirFn func() string
+
+	// execCmdFn can be set for tests to mock actual execution of the script. If
+	// nil, execCmd will be used, which has a different implementation on Windows
+	// and non-Windows platforms.
+	execCmdFn func(ctx context.Context, scriptPath string) ([]byte, int, error)
+}
+
+// Run processes all scripts identified by the execution IDs.
+func (r *Runner) Run(execIDs []string) error {
+	var errs []error
+
+	for _, execID := range execIDs {
+		if !r.ScriptExecutionEnabled {
+			if err := r.runOneDisabled(execID); err != nil {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		if err := r.runOne(execID); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		// NOTE: when we upgrade to Go1.20, we can use errors.Join, but for now we
+		// just concatenate the error messages in a single error that will be logged
+		// by orbit.
+		var sb strings.Builder
+		for i, e := range errs {
+			if i > 0 {
+				sb.WriteString("\n")
+			}
+			sb.WriteString(e.Error())
+		}
+		return errors.New(sb.String())
+	}
+	return nil
+}
+
+func (r *Runner) runOne(execID string) error {
+	const maxOutputRuneLen = 10000
+
+	script, err := r.Client.GetHostScript(execID)
+	if err != nil {
+		return fmt.Errorf("get host script: %w", err)
+	}
+
+	if script.ExitCode.Valid {
+		// already a result stored for this execution, skip, it shouldn't be sent
+		// again by Fleet.
+		return nil
+	}
+
+	runDir, err := r.createRunDir(execID)
+	if err != nil {
+		return fmt.Errorf("create run directory: %w", err)
+	}
+	// prevent destruction of dir if this env var is set
+	// TODO(mna): should we document this? Just an env var at the moment, not a cli flag.
+	if os.Getenv("FLEET_PREVENT_SCRIPT_TEMPDIR_DELETION") == "" {
```

Comment:
@georgekarrv Thoughts on documenting this or not? I added this env var to make it possible to keep the temporary directory of script execution (that gets deleted automatically after execution otherwise). As it stands I have not documented this.

Reply:
:+1: I'll do it in a separate PR for distinct approval/merging.